### PR TITLE
Fix vmazcopyread support in sockperf

### PR DESCRIPTION
**Pull request by MohammadQurt**
This fix sockperf support of zero copy  for TCP and UDP scenarios.

### CODE REVIEW

File: src/sockperf.cpp [hunk 1]
```diff
@@ -102,6 +102,9 @@ static bool sock_lib_started = 0; //
 static int s_fd_max = 0;
 static int s_fd_min = 0; /* used as THE fd when single mc group is given (RECVFROM blocked mode) */
 static int s_fd_num = 0;
+#ifdef USING_VMA_EXTRA_API
+static int s_fd_max_zcopy = 0;
```
Comment (MohammadQurt) [inline src/sockperf.cpp]: Didn't find a counter for TCP accepted sockets.
This counter will include TCP accepted sockets, and used for zcopy scenario only.

**Comment by igor-ivanov** [comment]
bot:retest

**Comment by MohammadQurt** [comment]
bot:retest

**Comment by igor-ivanov** [comment]
bot:retest

File: src/sockperf.cpp (hunk 1 above)
Comment (igor-ivanov) [reply src/sockperf.cpp]: I think you do not need this variable as far as `g_zeroCopyData` is map and you can get number of element using `g_zeroCopyData.size()`.

File: src/sockperf.cpp [hunk 2]
```diff
@@ -2093,7 +2096,7 @@ void cleanup() {
     }
 #ifdef USING_VMA_EXTRA_API
     if (g_vma_api && s_user_params.is_vmazcopyread) {
-        for (int i = s_fd_min; i < s_fd_max; i++) {
+        for (int i = s_fd_min; i <= s_fd_max_zcopy; i++) {
```
Comment (igor-ivanov) [inline src/sockperf.cpp]: You can do it w/o `for` cycle for map object. See https://www.geeksforgeeks.org/different-ways-delete-elements-stdmap-erase-clear/

File: src/sockperf.cpp [hunk 3]
```diff
@@ -3222,7 +3229,7 @@ void do_test() {
     info.fd_num = s_fd_num;
 #ifdef USING_VMA_EXTRA_API
     if (g_vma_api && s_user_params.is_vmazcopyread) {
-        for (int i = s_fd_min; i < s_fd_max; i++) {
+        for (int i = s_fd_min; i <= s_fd_max_zcopy; i++) {
```
Comment (igor-ivanov) [inline src/sockperf.cpp]: This code is useless as far as you allocate new object in `prepare_socket()`. So using this cycle you allocate memory for the same key twice.

File: src/server.h [hunk 4]
```diff
@@ -181,11 +184,15 @@ inline bool Server<IoType, SwitchActivityInfo, SwitchCalcGaps>::server_receive_t
...
+               !(nbytes = free_vma_packets(ifd, l_fds_ifd->recv.cur_size))) {
+        // Handled buffer is filled, free_vma_packets returns 0
+        ret = l_fds_ifd->recv.cur_size;
     } else
 #endif
     {
-        ret = msg_recvfrom(ifd, l_fds_ifd->recv.cur_addr + l_fds_ifd->recv.cur_offset,
-                           l_fds_ifd->recv.cur_size, &recvfrom_addr);
+        ret = msg_recvfrom(ifd, l_fds_ifd->recv.cur_addr + l_fds_ifd->recv.cur_offset, nbytes,
+                           &recvfrom_addr, &l_fds_ifd->recv.cur_addr);
```
Comment (igor-ivanov) [inline src/server.h]: no needs in changing `l_fds_ifd->recv.cur_size` with `nbytes`

File: src/client.h [hunk 5]
```diff
@@ -159,11 +162,15 @@ class Client : public ClientBase {
...
+                   !(nbytes = free_vma_packets(ifd, l_fds_ifd->recv.cur_size))) {
+            // Handled buffer is filled, free_vma_packets returns 0
+            ret = l_fds_ifd->recv.cur_size
         } else
 #endif
         {
-            ret = msg_recvfrom(ifd, l_fds_ifd->recv.cur_addr + l_fds_ifd->recv.cur_offset,
-                               l_fds_ifd->recv.cur_size, &recvfrom_addr);
+            ret = msg_recvfrom(ifd, l_fds_ifd->recv.cur_addr + l_fds_ifd->recv.cur_offset, nbytes,
+                               &recvfrom_addr, &l_fds_ifd->recv.cur_addr);
```
Comment (igor-ivanov) [inline src/client.h]: no needs in changing `l_fds_ifd->recv.cur_size` with `nbytes`

**Comment by igor-ivanov** [comment]
bot:retest

**Review by igor-ivanov — APPROVED** [review]
:+1

**Review by igor-ivanov — APPROVED** [review]
:+1:

**Comment by igor-ivanov** [comment]
merged as a part of https://github.com/Mellanox/sockperf/pull/116

**Pull request closed without merging**